### PR TITLE
Document how to deploy ZenML with custom Docker image

### PR DESCRIPTION
**Pull request by fa9r**
## Describe changes
Main docs page: https://app.gitbook.com/o/-MCl1Hlw9oU4xibZ-ymf/s/vlf6HVLGPRknfyMwMtR2/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Other (add details above)

### CODE REVIEW

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+---
+description: Deploying ZenML with custom Docker images.
+---
+
+In most cases, deploying ZenML with the default `zenmlhub/zenml-server` Docker 
+image should work just fine. However, there are some scenarios when you 
+might need to deploy ZenML with a custom Docker image:
+- You have implemented a custom artifact store for which you want to enable
+[artifact visualizations](../../../user-guide/advanced-guide/visualize-artifacts.md)
+or [step logs TODO LINK](../../../user-guide/advanced-guide/visualize-artifacts.md)
```
Comment (strickvl) [inline docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:10]: LINK

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+in your dashboard.
+- You have forked the ZenML repository and want to deploy a ZenML server based
+on your own fork because you made changes to the server / database logic.
+
+{% hint style="warning" %}
+Deploying ZenML with custom Docker images is only possible for Docker or Helm
+deployments.
```
Comment (strickvl) [inline docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:17]: Maybe link to the relevant docs for both of these?

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+{% endhint %}
+
+## Build and Push Custom ZenML Server Docker Image
+
+Here is how you can build a custom ZenML server Docker image:
+
+1. Set up a container registry of your choice. E.g., as an indivial developer
+you could create a free [dockerhub](https://hub.docker.com/) account and then 
+set up a free dockerhub repository.
```
Comment (strickvl) [inline docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:26]:
```suggestion
you could create a free [Docker Hub](https://hub.docker.com/) account and then 
set up a free Docker Hub repository.
```

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+See the general [ZenML CLI Deployment Guide](deploy-with-zenml-cli.md) for more
+information on how to use the `zenml deploy` CLI command and what other options
+can be configured.
+
+### Deploy a custom ZenML image via Docker
+
+To deploy your custom image via Docker, first familiarize yourself with the
+general [ZenML Docker Deployment Guide](deploy-with-docker.md).
+
+To use your own image, follow the general guide step-by-step but replace all
```
Comment (strickvl) [inline docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:97]:
```suggestion
To use your own image, follow the general guide step by step but replace all 
```

File: docs/book/user-guide/component-guide/artifact-stores/custom.md [hunk 5]
```diff
@@ -218,12 +218,11 @@ store. This means that your custom artifact store needs to be configured in a wa
...
-Furthermore, for deployed ZenML instances, you need to install the package dependencies of your artifact store
-implementation in the environment where you have deployed ZenML.
-
-You can do so either by building a custom Dockerfile based on
-the [ZenML base.Dockerfile](https://github.com/zenml-io/zenml/blob/main/docker/base.Dockerfile) and using that to
-redeploy your ZenML instance or by installing the respective packages in the deployed environment manually.
+Furthermore, for deployed ZenML instances, you need to install the package 
+dependencies of your artifact store implementation in the environment where you 
+have deployed ZenML. See the
+[Docs on deploying ZenML with custom Docker images](../../../platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md)
```
Comment (strickvl) [inline docs/book/user-guide/component-guide/artifact-stores/custom.md:224]:
```suggestion
[documentation on deploying ZenML with custom Docker images](../../../platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md)
```

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md (hunk 1 above)
Comment (fa9r) [reply docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:10]: Yep this needs to wait till the step logs page exists (which @bcdurak will write this sprint)

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md (hunk 2 above)
Comment (fa9r) [reply docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:17]: I have links for both below in the respective sections, you think it's useful to add them here as well?

Comment (strickvl) [reply docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:17]: Yes. Links are free + while reading I don't necessarily want to keep reading when maybe I just want to go there first.

**Review by schustmi — CHANGES_REQUESTED** [review]
For the deployment using CLI/helm, do we need an additional section on `imagePullSecrets` so that the (potentially) private image that they pushed can be pulled?

File: docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md (hunk 1 above)
Comment (fa9r) [reply docs/book/platform-guide/set-up-your-mlops-platform/deploy-zenml/deploy-with-custom-image.md:10]: Actually I just checked and we don't have a ticket for that, we only have a ticket to document how the logs can be switched off which will not become a separate section I assume.

I linked it to best fitting section we have for now.

**Comment by fa9r** [comment]
> For the deployment using CLI/helm, do we need an additional section on imagePullSecrets so that the (potentially) private image that they pushed can be pulled?

Seems like no one knows how exactly this would need to be done so I'd say let's wait until some user actually needs this before investing time into researching it

**Review by schustmi — APPROVED** [review]

**Review by strickvl — APPROVED** [review]

**Pull request merged**